Source/NSTextFinder.m

```diff
@@ -0,0 +1,333 @@
+                        forward: NO];
+}
+
+- (void) _replaceAll
+{
+  [_finder replaceAllInTextView: (NSTextView *)_client
+                onlyInSelection: NO];
+}
+
+- (void) _replace
+{
+  [_finder replaceStringInTextView: (NSTextView *)_client];
+}
+
+- (void) _replaceAndFind
+{
+   [_finder replaceStringInTextView: (NSTextView *)_client];
+   [_finder findStringInTextView: (NSTextView *)_client
+                         forward: YES];
+}
+
+- (void) _setSearchString
+{
+  [_finder takeFindStringFromTextView: (NSTextView *)_client];
+}
+
+- (void) _replaceAllInSelection
+{
+  [_finder replaceAllInTextView: (NSTextView *)_client
+                onlyInSelection: YES];
+}
+
+- (void) _selectAll
+{
+}
+
+- (void) _selectAllInSelection
+{
+}
+
+- (void) _hideFindInterface
+{
+}
+
+- (void) _showReplaceInterface
+{
+  [_finder showFindPanel];
+}
+
+- (void) _hideReplaceInterface
+{
+}
+
+// Validating and performing
+- (void) performTextFinderAction: (id)sender
+{
+  [self performFindPanelAction: sender];
+}
+
+- (void) performFindPanelAction: (id)sender
+{
+  if (_finder == nil)
+    {
+      _finder = [[GSTextFinder alloc] init];
+    }
+  [self performAction: [sender tag]];
+  [self validateUserInterfaceAction: sender];
+}
+
+- (void) performAction: (NSTextFinderAction)op
+{
+  switch (op)
+    {
+    case NSTextFinderActionShowFindInterface:
+      [self _showFindInterface];
+      break;
+    case NSTextFinderActionNextMatch:
+      [self _nextMatch];
+      break;
+    case NSTextFinderActionPreviousMatch:
+      [self _previousMatch];
+      break;
+    case NSTextFinderActionReplaceAll:
+      [self _replaceAll];
+      break;
+    case NSTextFinderActionReplace:
+      [self _replace];
+      break;
+    case NSTextFinderActionReplaceAndFind:
+      [self _replaceAndFind];
+      break;
+    case NSTextFinderActionSetSearchString:
+      [self _setSearchString];
+      break;
+    case NSTextFinderActionReplaceAllInSelection:
+      [self _replaceAllInSelection];
+      break;
+    case NSTextFinderActionSelectAll:
+      [self _selectAll];
+      break;
+    case NSTextFinderActionSelectAllInSelection:
+      [self _selectAllInSelection];
+      break;
+    case NSTextFinderActionHideFindInterface:
+      [self _hideFindInterface];
+      break;
+    case NSTextFinderActionShowReplaceInterface:
+      [self _showReplaceInterface];
+      break;
+    case NSTextFinderActionHideReplaceInterface:
+      [self _hideReplaceInterface];
+      break;
+    default:
+      NSLog(@"Unknown operation: %ld", op);
+      break;
+    }
+}
+
+- (BOOL) validateUserInterfaceAction: (id<NSValidatedUserInterfaceItem>)item
+{
+  SEL action = [item action];
+  if (sel_isEqual(action, @selector(performTextFinderAction:)) ||
+      sel_isEqual(action, @selector(performFindPanelAction:)))
+    {
+      return [self validateAction: [item tag]];
+    }
+
+  return YES;
+}
+
+- (BOOL) validateAction: (NSTextFinderAction)op
+{
+  return YES;
```

Comment:
The actual code is missing here. See GSTextFinder for an example.

Reply:
Done.